src/coreclr/vm/ceemain.cpp

```diff
@@ -1671,6 +1677,26 @@ BOOL STDMETHODCALLTYPE EEDllMain( // TRUE on success, FALSE on error.
                 }
                 break;
             }
+
+            case DLL_THREAD_DETACH:
+            {
+#ifdef TARGET_WINDOWS
+                // Make sure that we do not have an attached Thread object.
+                // We can end up here with live Thread if some kind of thread termination callback initializes
+                // the Thread by entering managed code or by calling APIs that set up Thread, after the point where
+                // the OS premortem callback for the thread could have run.
+                // There will be no further clean up at this point and we will likely crash in GC once OS clears the native TLS.
+                //
+                // NB: It is ok for the Thread to be reinitialized before the premortem OS callback runs.
+                //     That historically may happen in rare cases when a thread cleans upon returning, but then FlsDataCleanup for COM
+                //     ends up releasing objects and calling APIs in the runtime that require Thread, or sends messages to the thread's
+                //     managed message loop. That is ok, as long as initialization happens prior to our thread termination
+                //     callback from OS. We guarantee that by ensuring our FlsSlot is initializad after COM has already been
+                //     initialized.
+                _ASSERTE_ALL_BUILDS(!GetThreadNULLOk() && "Thread initialized after final clean up could have run.");
+#endif
+                break;
+            }
```

Comment:
```suggestion
```
This should not be needed with the other change. It would be best to get rid of this DllMain in a follow up change.

Reply:
This can be hit if a thread did not run any managed code at all, and then someone enters managed code in an FLS callback that runs after our FLS callback would have run. 

We do not know any cases like this right now. It is theoretically possible.

Reply:
I guess we could drop this check, considering we do not like to have this DllMain

Reply:
You made a good point that this check can catch one of the corner cases that I have not considered. I think that it is very unlikely for user code to hit this corner case, without tripping over the other check most of the time.

I am fine with either keeping it or leaving it out.